tests/unit/test_characterProcessing.py

```diff
@@ -0,0 +1,119 @@
+# A part of NonVisual Desktop Access (NVDA)
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+# Copyright (C) 2020 NV Access Limited
+
+"""Unit tests for the characterProcessing module.
+"""
+
+import unittest
+import re
+from characterProcessing import SYMLVL_ALL
+from characterProcessing import SpeechSymbolProcessor
+from characterProcessing import processSpeechSymbols as process
+
+
+class TestComplex(unittest.TestCase):
+	"""Test the complex symbols rules.
+	"""
+
+	def _replace_cb(self, name, replacement):
+		"""Return a regexp callback which replaces matches of the given
+		group name with the replacement string, with support for
+		replacement of group references.
+		"""
+		def replace(m):
+			if m.lastgroup == name:
+				return SpeechSymbolProcessor._replaceGroups(self, m, replacement)
+			return m.group()
+		return replace
+
+	def _replace(self, string, pattern, name, replacement):
+		"""Perform a pattern-based replacement on a string, for one
+		named group, with support for replacement of group references.
+		"""
+		regexp = re.compile(pattern, re.UNICODE)
+		return regexp.sub(self._replace_cb(name, replacement), string)
+
+	def test_group_replacement(self):
+		"""Test that plain text gets properly replaced
+		"""
+		replaced = self._replace(
+			string="1",
+			pattern=r"(?P<foo>(\d))",
+			name="foo",
+			replacement="a"
+		)
+		self.assertEqual(replaced, "a")
+
+	def test_backslash_replacement(self):
+		"""Test that backslashes get properly replaced
+		"""
+		replaced = self._replace(
+			string="1",
+			pattern=r"(?P<foo>(\d))",
+			name="foo",
+			replacement=r"\\"
+		)
+		self.assertEqual(replaced, "\\")
+
+	def test_double_backslash_replacement(self):
+		"""Test that double backslashes get properly replaced
+		"""
+		replaced = self._replace(
+			string="1",
+			pattern=r"(?P<foo>(\d))",
+			name="foo",
+			replacement=r"\\\\"
+		)
+		self.assertEqual(replaced, r"\\")
+
+	def test_unknown_escape(self):
+		"""Test that a non-supported escaped character raises an error
+		"""
+		with self.assertRaises(LookupError):
+			self._replace(
+				string="1",
+				pattern=r"(?P<foo>(\d))",
+				name="foo",
+				replacement=r"\a"
+			)
+
+	def test_unterminated_escape(self):
+		"""Test that an escape at the end of string raises an error
+		"""
+		with self.assertRaises(LookupError):
+			self._replace(
+				string="1",
+				pattern=r"(?P<foo>(\d))",
+				name="foo",
+				replacement="\\"
+			)
+
+	def test_group_replacements(self):
+		"""Test that group references get properly replaced
+		"""
+		replaced = self._replace(
+			string="bar.BAT",
+			pattern=r"(?P<foo>([a-z]*)\.([A-Z]*))",
+			name="foo",

```

Comment:
I've just noticed that all these tests use name, when it is not strictly necessary to test the replace function. Why is this? I want to see tests for the custom code in the `_replaceGroups` code, which doesn't consider group name at all. The more simple the regex, the easier these tests will be to understand.

Reply:
Right, we can just put anonymous groups. The only exception if of course the test for multiple groups, we want to name the one we aim to replace.
Done so.